### PR TITLE
Drive data callbacks on the output write event instead of the input read event

### PR DESCRIPTION
**Pull request by padenot**
This fixes the output buffering we observe when doing calls in Firefox.

### CODE REVIEW

**Comment by padenot** [comment]
@kinetiknz @ChunMinChang @achronop, it looks like I can't ask for a review here, but please have a look.

**Review by achronop — APPROVED** [review]
It looks sensible. I have no comments so r+ from me. Better to get a second review, just in case, mostly about the rust details.

File: src/backend/stream.rs
Comment (ChunMinChang) [inline src/backend/stream.rs]: It would be better if we can keep the `Debug` trait. If the newly added `enum`s are not derivable, we can implement them explicitly. Something like:
```rust
struct NoDebug;

enum NonDerivable {
    Foo(NoDebug)
}

impl std::fmt::Debug for NonDerivable {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            &NonDerivable::Foo(_) => write!(f, "Foo"),
        }
    }
}
```

File: src/backend/stream.rs
Comment (ChunMinChang) [inline src/backend/stream.rs:142]: Is it possible to merge these three `enum` into one, some kind of `BufferManager<T>`? It would make sure we use the same type for `Consumer`, `Producer` and `InputBuffer`

File: src/backend/stream.rs
Comment (ChunMinChang) [inline src/backend/stream.rs:131]: `RingBufferConsumer::IntegerRingBufferConsumer` or `RingBufferConsumer::FloatRingBufferConsumer`is a bit long.

File: src/backend/stream.rs
Comment (ChunMinChang) [inline src/backend/stream.rs]: `/* On output only case need to ....` not output only anymore? If so, we should update the comment here.

File: src/backend/stream.rs
Comment (ChunMinChang) [inline src/backend/stream.rs:371]: Can we always read the data pushed into the buffer? What is the relationship between `nsamples` and `read_samples` in `read_data`? I guess the `sample_size()` of `output_sample_spec` and `input_sample_spec` are same since they are validated in [cubeb.c](https://github.com/kinetiknz/cubeb/blob/0f799155fc1689c4932589b0240b604374419615/src/cubeb.c#L91). Thus, their relationship should depend on `nbytes` with `read_size`  in `read_data`, but I am not familiar with them.

**Review by ChunMinChang — COMMENTED** [review]
Overall looks good. 

It would be easier to check the type we use if those added `enum`s can be merged into one struct with the same type.

File: src/backend/stream.rs
Comment (kinetiknz) [reply src/backend/stream.rs:131]: You could `use RingBufferConsumer::*` to avoid repeating the enum name everywhere.

File: src/backend/stream.rs
Comment (kinetiknz) [inline src/backend/stream.rs]: This log message seems incomplete?

File: src/backend/stream.rs
Comment (kinetiknz) [inline src/backend/stream.rs]: Typo: underrunning

Should the error cases log something?

File: Cargo.toml
Comment (kinetiknz) [inline Cargo.toml:18]: License is okay (MIT/Apache2).  Uses SeqCst everywhere, so performance is probably not ideal.

File: src/backend/stream.rs
Comment (kinetiknz) [inline src/backend/stream.rs]: samples

File: src/backend/stream.rs
Comment (kinetiknz) [inline src/backend/stream.rs]: users -> callee

File: src/backend/stream.rs
Comment (kinetiknz) [inline src/backend/stream.rs]: Add an assert or a comment where tlength/fragsize are set to connect these together.

**Review by kinetiknz — APPROVED** [review]
It would be good to have an explanation of what specifically was wrong with the old input-driven approach.  It's not evident from these changes.

File: src/backend/stream.rs
Comment (padenot) [reply src/backend/stream.rs]: I don't know of a way to avoid repetition. Maybe a macro?

File: src/backend/stream.rs
Comment (padenot) [reply src/backend/stream.rs]: Fixed, thanks.

File: src/backend/stream.rs
Comment (padenot) [reply src/backend/stream.rs:131]: Yep, thanks.

File: src/backend/stream.rs
Comment (padenot) [reply src/backend/stream.rs:142]: Much cleaner indeed, thanks.

Comment (padenot) [reply src/backend/stream.rs:142]: Not generic of course, otherwise we just push the problem, for now, we need to dynamically dispatch.

File: src/backend/stream.rs
Comment (padenot) [reply src/backend/stream.rs:371]: It depends on how the streams configured. `sample_size()` is guaranteed to be the same for input and output (it's either 2 for int16 or 4 for float32).

`nbytes`, `read_size`, `read_data` depend on how the `pa_buffer_attr` has been set. Because it's real time, and the sample-rates and latencies are the same, you have the same number of input frames that you have output frames to produce, so it generally balances out. Worst case we add silence to compensate.

File: src/backend/stream.rs
Comment (padenot) [reply src/backend/stream.rs]: Yes thanks. It looks like this ringbuffer does not have a way to get the number of elements inside it.

File: Cargo.toml
Comment (padenot) [reply Cargo.toml:18]: Yes, I'll get use something better based off the cubeb/gecko one (that I have optimized a bit locally to avoid false sharing), this is a bit crude. It should be straightforward to port it to rust, or to fix `ringbuf` to do the same.

We'll do something similar in some cases for the AudioUnit backend, when we can't use aggregate devices because the input and output stream are not on the same physical device.

**Comment by padenot** [comment]
> It would be good to have an explanation of what specifically was wrong with the old input-driven approach. It's not evident from these changes.

It's a bit unclear. Ultimately, it's a bit weird to drive a playback mechanism based on the speed at which some input audio is pushed, though. I've been trying to ping people that know pulse a bit better to understand but they haven't answered yet.

Another data point is that we're the only project doing duplex stream based on the input callbacks, and the only project that has problems.

File: src/backend/stream.rs
Comment (padenot) [reply src/backend/stream.rs]: I've done it on the `BufferManager`, it's not too verbose.

**Comment by achronop** [comment]
I put here, as a reference, the discussion I had with PA people when I made that change. Some answers may be lurking there.

https://www.mail-archive.com/pulseaudio-discuss@lists.freedesktop.org/msg14786.html

**Comment by padenot** [comment]
I can't merge it seems, but I'd like to have this one quickly in Gecko. This has been a pain for users for quite a long time.

File: src/backend/stream.rs
Comment (ChunMinChang) [reply src/backend/stream.rs:142]: Using a trait object with `Box` should work for dynamical dispatch(something like [this](https://searchfox.org/mozilla-central/rev/da855d65d1fbdd714190cab2c46130f7422f3699/media/libcubeb/cubeb-coreaudio-rs/src/backend/auto_array.rs#6)). A better approach is to define the type(`f32` or `i16`) when initializing the stream. It's the same issue we discussed in the Rust AudioUnit backend. But we can refactor it later.

**Comment by kinetiknz** [comment]
> I can't merge it seems, but I'd like to have this one quickly in Gecko. This has been a pain for users for quite a long time.

The review comments haven't been addressed.  Did you forget to push the updated changeset?

File: src/backend/stream.rs
Comment (kinetiknz) [reply src/backend/stream.rs]: We're not removing Debug from PulseStream.

**Comment by padenot** [comment]
wtf this is the the right commit at all and git says `Everything up-to-date`. Thanks for the catch @kinetik.

**Comment by padenot** [comment]
Yeah no I was just dumb and didn't `git add `. Too much `hg` I suppose, where it's not necessary. This is the right commits, with all comments from both of you addressed, sorry about that.

**Review by kinetiknz — APPROVED** [review]
LGTM, thanks.

I was confused when I read the bug, sorry, I didn't click that the new commit was missing until I saw your reference to BufferManager.

**Comment by kinetiknz** [comment]
CI is failing because ringbuf requires a newer rustc.  Reverting PR #41 dropped us back to 1.21, but the Travis config is out of date for Gecko's minimum.

**Pull request merged**
